src/communication/makeRetrying.ts

```diff
@@ -0,0 +1,115 @@
+  const retryAfter = parseInt(response.headers['retry-after'], 10);
+  if (isNaN(retryAfter)) {
+    return undefined;
+  }
+  // Convert the header value from seconds to milliseconds.
+  return retryAfter * 1e3;
+}
+
+/**
+ * Enhances the passed Axios instance, making it attempt requests multiple times in some scenarios.
+ *
+ * The idea is that if the Mollie API has a brief hiccup, the extra attempts may cause the request to succeed anyway.
+ *
+ * If the Mollie API responds with a 5×× status code, the request will be re-attempted until:
+ *  * the Mollie API responds with a different status code, or
+ *  * the attempt limit has been reached (it gives up after the third attempt), or
+ *  * the request has timed out (as per the timeout set in the Axios instance).
+ *
+ * For `POST` and `DELETE` requests, an idempotency key is added. This ensures the Mollie API can distinguish a single
+ * request being re-attempted from two separate similarly-looking requests. In effect, this allows this client to
+ * safely re-attempt requests.
+ */
+export default function makeRetrying(axiosInstance: AxiosInstance) {
+  // Set the attempt (in the request configuration) of any request.
+  axiosInstance.interceptors.request.use((config: AxiosRequestConfig & Partial<AttemptState>) => {
+    config[attemptIndex] = (config[attemptIndex] ?? -1) + 1;
+    return config;
+  });
+  // Intercept any erroneous responses, and consider doing another attempt.
+  axiosInstance.interceptors.response.use(undefined, error => {
+    // If the request configuration is unexpected ‒ in other words, it seems the request did not go through the
+    // interceptor above ‒ do not make another attempt.
+    if (!checkError(error)) {
+      return Promise.reject(error);
+    }
+    const { config } = error;
+    // If the attempt limit has been reached, do not make another attempt.
+    if (config[attemptIndex] == attemptLimit - 1) {
+      return Promise.reject(error);
+    }
+    // If there is no response or the HTTP status code of the response is not 5××, do not make another attempt.
+    if (error.response == undefined || Math.floor(error.response.status / 100) != 5) {
+      return Promise.reject(error);
+    }
+    // Determine the delay after which the next attempt is made, preferring a Retry-After header defined in the
+    // response over the default value.
+    const delay = parseRetryAfterHeader(error.response) ?? retryDelay;
+    // Schedule the attempt.
+    return new Promise(resolve => setTimeout(() => resolve(axiosInstance.request(config)), delay));
+  });
+  // Overwrite the post and delete functions in the Axios instance, making them add the header. Note that the Axios
+  // instance still has its original request function, which is not overwritten. Additionally, the Axios instance
+  // itself can be called as a function, which is equivalent to calling the request function. In this client, neither
```

Comment:
Is this a ticking time bomb? Or do we think this is okay, because worst case a call won't be retried?

Reply:
A ticking time bomb in the sense that it explodes if someone in the future does `axiosInstance.request` instead of `axiosInstance.post` or `axiosInstance.delete`? Then yes.

The worst case scenario would be bad in that case. It _will_ still retry (because the interceptors are still in effect) but it _won't_ add the idempotency key header. This means any `POST`s or `DELETE`s would be unsafe (as the Mollie API wouldn't be able to distinguish retries from separate requests).

On the bright side: our test suite does specifically check that the header is set.

What are your thoughts after hearing this?

Reply:
I'm not so concerned that consumers of the library will use the axios instance directly. I'm more concerned that someone adds functionality to this library, and uses `axiosInstance.request(...)`. How hard would it be to intercept those calls as well?

Reply:
We're not exposing the Axios instance. (It is technically available as `mollieClient.payments.networkClient.networkClient.axiosInstance`, but using it means ignoring all documentation and type definitions, and is not realistic.)

We could override `axiosInstance.request` instead of `axiosInstance.post` and `axiosInstance.delete`, as those use `.request` internally. However, the Axios instance itself is also callable: `axiosInstance(…)`. We would have to look into overriding the behaviour there as well.

Reply:
Right. Well with that backdrop, my gut feeling is that there is some merit in overriding `.request()` instead of `.post()` and `.delete()`. I can realistically imagine a future contributor using `.request()` instead of `.post()`.

It's unfortunate that it's difficult to override calling the instance, because there's actually no reason that `.request()` is more likely to be used than `axiosInstance(...)`. So we would definitely want some documentation or comments on this.

Reply:
We've changed the implementation to use an interceptor instead of overriding any functions, defusing your time bomb.